From 7378b89643a3bdc3744a6144f0896860afeb8cd5 Mon Sep 17 00:00:00 2001
From: Theo GOUREAU <theo.goureau-ext@st.com>
Date: Tue, 7 Jan 2025 14:13:32 +0100
Subject: [PATCH 764/954] feat(stm32mp2): workaround implementation for D1/D2
 low power M0+ crash

This commit implements a workaround for the D1/D2 low power M0+ crash.
The workaround is implemented in the M0+ firmware, and this patch
triggers it by letting the A35 send an IRQ to the M0+ via EXTI2 line
60 when the last CPU running is about to enter low power.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git ec64537eecfea48ad908a9f3ebac342b7ea25985 feat(stm32mp2): workaround implementation for D1/D2 low power M0+ crash]
Change-Id: Ieb88a7842cc2bc64cd02beed3529225580a74cac
Signed-off-by: Theo GOUREAU <theo.goureau-ext@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/427332
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 plat/st/stm32mp2/stm32mp2_def.h |  3 +++
 plat/st/stm32mp2/stm32mp2_pm.c  | 20 ++++++++++++++++++++
 2 files changed, 23 insertions(+)

diff --git a/plat/st/stm32mp2/stm32mp2_def.h b/plat/st/stm32mp2/stm32mp2_def.h
index 447dbef2f3..19cf8c9c0b 100644
--- a/plat/st/stm32mp2/stm32mp2_def.h
+++ b/plat/st/stm32mp2/stm32mp2_def.h
@@ -370,6 +370,7 @@ enum ddr_type {
 #define STM32MP_EXTI2_BASE		U(0x442D0000)
 #endif /* !STM32MP21 */
 
+#define EXTI2_SWIER2			U(0x28)
 #define EXTI2_C1IMR1			U(0x80)
 #define EXTI2_C1IMR2			U(0x90)
 #if !STM32MP21
@@ -385,6 +386,8 @@ enum ddr_type {
 #endif /* !STM32MP21 */
 
 #define EXTI_CmCIDCFGR(n)		(0x300U + ((n) * 4U))
+#define EXTI2_BIT(_id)			BIT_32(_id % 32)
+#define EXTI2_C1SEV			EXTI2_BIT(60)
 
 /*******************************************************************************
  * STM32MP2 GPIO
diff --git a/plat/st/stm32mp2/stm32mp2_pm.c b/plat/st/stm32mp2/stm32mp2_pm.c
index 677ad16d66..3138c77ed3 100644
--- a/plat/st/stm32mp2/stm32mp2_pm.c
+++ b/plat/st/stm32mp2/stm32mp2_pm.c
@@ -469,6 +469,17 @@ bool stm32_pwr_cpu2_state_is_running(uintptr_t pwr_base)
 #endif
 }
 
+#if !STM32MP_M33_TDCID
+static bool stm32_pwr_cpu3_state_is_running(__maybe_unused uintptr_t pwr_base)
+{
+#if STM32MP21
+	return false;
+#else
+	return (mmio_read_32(pwr_base + PWR_CPU3D3SR) & PWR_CPU3D3SR_CSTATE_MASK) != 0U;
+#endif
+}
+#endif
+
 static int stm32_pwr_domain_validate_suspend(const psci_power_state_t *target_state)
 {
 	uintptr_t pwr_base = stm32mp_pwr_base();
@@ -827,6 +838,7 @@ static void __dead2 stm32_pwr_domain_pwr_down_wfi(const psci_power_state_t
 						  *target_state)
 {
 	u_register_t mpidr = read_mpidr();
+	uintptr_t pwr_base __maybe_unused = stm32mp_pwr_base();
 	unsigned int core_id = MPIDR_AFFLVL0_VAL(mpidr);
 
 	/* Core is no more running (stopped or suspended) */
@@ -860,6 +872,14 @@ static void __dead2 stm32_pwr_domain_pwr_down_wfi(const psci_power_state_t
 		stm32_pm_context_save(target_state);
 	}
 
+#if !STM32MP_M33_TDCID
+	bool is_cpu3_running = stm32_pwr_cpu3_state_is_running(pwr_base);
+	if (psci_is_last_on_cpu_safe() && is_cpu3_running) {
+		/* Send an IRQ to the M0+ using EXTI2 C1SEV to warn about D1/D2 standby. */
+		mmio_write_32(STM32MP_EXTI2_BASE + EXTI2_SWIER2, EXTI2_C1SEV);
+        }
+#endif /* !STM32MP_M33_TDCID */
+
 	/* Synchronize instruction flow before auto-reset from WFI */
 	isb();
 	psci_power_down_wfi();
-- 
2.53.0

